MudAutoComplete: Add BeforeItemsTemplate and AfterItemsTemplate

derekwelton opened the pull request:
## Description
This adds an additional render fragment for use cases when you want to always show something at the end of the pop up. An example of this is adding a button at the end to add additional items to the list no matter how many items there are in the list.

![image](https://user-images.githubusercontent.com/31353783/234706374-178183c5-0b11-406a-85d0-c0ec38feee65.png)


## How Has This Been Tested?
No new test are required. All that was added is a render fragment.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)



## Checklist:
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

derekwelton committed:
MudAutoComplete: Add EndItems Render Fragment

Optional presentation template for when you need to section to always show in the drop down

ScarletKuro commented:
> No new test are required. All that was added is a render fragment.

We need test coverage of the new feature. A simple doc example would be nice to have also.
The purpose is to check in case that if someone refactors the file that the feature is still in place and working.

Mr-Technician commented:
I like this but fear it _could_ be bloating the API. Thoughts @henon?

henon commented:
It is a legit use case, I can also think of another one like "Show more results" which will increase the number of shown hits and re-filter.

henon commented:
Maybe we add a corresponding template for the start of the list as well? I am thinking about good naming, so maybe `ListStart` and `ListEnd`? We should have done this instead of a specialized `MoreItemsTemplate` but it is too late to change that now.

derekwelton commented:
> Maybe we add a corresponding template for the start of the list as well? I am thinking about good naming, so maybe `ListStart` and `ListEnd`? We should have done this instead of a specialized `MoreItemsTemplate` but it is too late to change that now.

I agree with the naming. I named it the way I did to follow the convention of "MoreItemsTemplate" and "NoItemsTemplate". I like the idea of adding a "Start" one as well.

Mr-Technician commented:
> We should have done this instead of a specialized `MoreItemsTemplate` but it is too late to change that now.

I don't think so because `MoreItemsTemplate` is distinct from something that is always shown after the items list. There is specific logic that determines when it shows.

derekwelton committed:
MudAutoComplete: Created ListStart RenderFragment and Test

Created another render fragment at the start of the popup, renamed EndItems to ListEnd, added test as well

derekwelton commented:
I created the Render Fragment for the ListStart and added some test as well. Once we agree on same names, I can add another commit with the name changes. Currently the names are: `ListEndTemplate` and `ListStartTemplate`

derekwelton committed:
Attempted To Fix Test Issues

ScarletKuro commented on src/MudBlazor.sln:
```diff
@@ -27,7 +27,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MudBlazor.Examples.Data", "
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MudBlazor.Docs.Server", "MudBlazor.Docs.Server\MudBlazor.Docs.Server.csproj", "{90B6D51A-133A-4744-A5FE-FF398F0D6BAA}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "MudBlazor.SourceCodeGenerator", "MudBlazor.SourceCodeGenerator\MudBlazor.SourceCodeGenerator.csproj", "{CDE02174-96AB-4E97-AFA8-391C0FF98225}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MudBlazor.SourceCodeGenerator", "MudBlazor.SourceCodeGenerator\MudBlazor.SourceCodeGenerator.csproj", "{CDE02174-96AB-4E97-AFA8-391C0FF98225}"
```
This file change should be reverted

derekwelton replied:
Could Catch, thanks!

derekwelton committed:
Fixed Autocomplete_Should_LoadListStartWhenSet Test

derekwelton committed:
Reverted SLN Change

henon commented:
> Also, maybe `BeforeItemsTemplate` and `AfterItemsTemplate` would be appropriate names?

Those are good names, I agree!

derekwelton committed:
Renamed ListStart and ListEnd

derekwelton commented:
I've renamed the Render Fragments as requested. We should be all set unless I missed something.

henon commented:
Can you add a sentence to the Presentation section of the docs, similar to the one I underlined?
![image](https://user-images.githubusercontent.com/44090/234926256-6ef68954-afe4-49ab-9525-7b56b68a38d6.png)

I think we don't need to complicate the example though, the description should be enough.

derekwelton committed:
Added Documentation

henon commented on src/MudBlazor.Docs/Pages/Components/Autocomplete/AutocompletePage.razor:
```diff
@@ -51,7 +51,8 @@
             <SectionHeader Title="Presentation Extras">
                 <Description>
-                    You can also define a <CodeInline Tag="true">MoreItemsTemplate</CodeInline> or a <CodeInline Tag="true">NoItemsTemplate</CodeInline> for special scenarios.
+                    You can also define a <CodeInline Tag="true">MoreItemsTemplate</CodeInline>, <CodeInline Tag="true">NoItemsTemplate</CodeInline>, 
+                    <CodeInline Tag="true">BeforeItemsTemplate</CodeInline>, or <CodeInline Tag="true">AfterItemsTemplate</CodeInline>  for special scenarios.
```
We can do better than that :). How about this:

You can also define a `<CodeInline Tag="true">MoreItemsTemplate</CodeInline>` to handle the case where the search returns more items than the list is configured to show and a `<CodeInline Tag="true">NoItemsTemplate</CodeInline>` that presents some default content if the search returns no results. For your own custom content that always shows at the top or the bottom of the list define the                     `<CodeInline Tag="true">BeforeItemsTemplate</CodeInline>` and the `<CodeInline Tag="true">AfterItemsTemplate</CodeInline>`.
```

derekwelton replied:
Done :)

henon commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor:
```diff
@@ -38,10 +38,16 @@
                 <MudPopover Open="@IsOpen" MaxHeight="@MaxHeight" AnchorOrigin="@AnchorOrigin" TransformOrigin="@TransformOrigin" Class="@PopoverClass" RelativeWidth="true">
-                    @if(ProgressIndicatorInPopoverTemplate != null && IsLoading)
+                    @if (BeforeItemsTemplate != null)
+                    {
+                        <div class="pa-1">
+                            @BeforeItemsTemplate
+                        </div>
+                    }
+                    @if (ProgressIndicatorInPopoverTemplate != null && IsLoading)
                     {
                         @ProgressIndicatorInPopoverTemplate
-                    } 
+                    }
```
We need to think if the ProgressIndicatorTemplate shouldn't be on top of the BeforeItemsTemplate. What do you think @Mr-Technician ?

ScarletKuro replied:
In my opinion loading indicator should be always on top.

derekwelton replied:
Do we even show the `BeforeItemsTemplate` and `AfterItemsTemplate` if `IsLoading` is true? I think we should not have them render if its true. This makes sense to me.

Mr-Technician replied:
Yes to both @ScarletKuro and @derekwelton.

derekwelton replied:
Changes have been made.

derekwelton committed:
Additional Modifications to Docs

Mr-Technician commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -263,6 +263,21 @@ public partial class MudAutocomplete<T> : MudBaseInput<T>, IDisposable
 
+
+        /// <summary>
+        /// Optional presentation template for when you need to section to always show at the start in the drop down
```
There is a grammatical error in this sentence. I would probably say "Optional presentation template that is always shown at the top of the list"

derekwelton replied:
@Mr-Technician Resolved

Mr-Technician commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -263,6 +263,21 @@ public partial class MudAutocomplete<T> : MudBaseInput<T>, IDisposable
+        public RenderFragment BeforeItemsTemplate { get; set; }
+
+        /// <summary>
+        /// Optional presentation template for when you need to section to always show at the end in the drop down
```
"Optional presentation template that is always shown at the bottom of the list"

derekwelton replied:
@Mr-Technician Resolved

derekwelton committed:
Grammatical Fixes && Positioning improvemnts for BeforeItems Template

derekwelton committed:
Attempted to Fix Autocomplete_Should_LoadListStartWhenSet Test

derekwelton committed:
Fixed Issue with test not passing.

derekwelton committed:
Cleaned Up Test

henon commented:
Thanks @derekwelton!

rf-0 commented:
This will be useful to show recent/previous searches when selecting the autocomplete box without any text input (first click).
